mainnet_contract/contracts/HeaderReader.sol

```diff
@@ -0,0 +1,423 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.4.21 <0.9.0;
+pragma experimental ABIEncoderV2;
+
+library HeaderReader {
+
+  uint8 constant STRING_SHORT_START = 0x80;
+  uint8 constant STRING_LONG_START = 0xb8;
+  uint8 constant LIST_SHORT_START = 0xc0;
+  uint8 constant LIST_LONG_START = 0xf8;
+  uint8 constant WORD_SIZE = 32;
+
+  struct RLPItem {
+    uint256 len;
+    uint256 memPtr;
+  }
+
+  function getParentHashAndNumber(bytes memory genesis_header) public pure returns (bytes32, int) {
+    RLPItem[] memory ls = toList(toRlpItem(genesis_header));
+    return (toBytes32(toBytes(ls[0])), int(toUint(ls[8])));
+  }
+
+  // Returns (parentHash, number, round_number, certificate hash, sigs)
+  function getValidationParams(bytes memory header) public pure returns (bytes32, int, uint64, bytes32, bytes[] memory) {
+    RLPItem[] memory ls = toList(toRlpItem(header));
+    RLPItem[] memory extra = toList(toRlpItem(getExtraData(toBytes(ls[12]))));
+    uint64 round_number = uint64(toUint(extra[0]));
+    RLPItem[] memory raw_sigs = toList(toList(extra[1])[1]);
+    bytes[] memory sigs = new bytes[](raw_sigs.length);
+    for (uint i = 0; i < raw_sigs.length; i++) {
+      sigs[i] = toBytes(raw_sigs[i]);
+    }
+    bytes32 signHash = createSignHash(toBytes32(toBytes(ls[0])), round_number, int(toUint(ls[8])), uint64(toUint(toList(extra[1])[2])));
+    return (toBytes32(toBytes(ls[0])), int(toUint(ls[8])), round_number, signHash, sigs);
+  }
+
+  function getExtraData(bytes memory extra) public pure returns (bytes memory) {
+    bytes memory extraData = new bytes(extra.length-1);
+    uint extraDataPtr;
+    uint extraPtr;
+    assembly { extraDataPtr := add(extraData, 0x20) }
+    assembly { extraPtr := add(extra, 0x21) }
+    copy(extraPtr, extraDataPtr, extra.length-1);
+    return extraData;
+  }
+
+  function createSignHash(bytes32 block_hash, uint64 round_num, int number, uint64 gap_num) internal pure returns (bytes32 signHash) {
+    bytes[] memory x = new bytes[](3);
+    x[0] = encodeBytes(abi.encodePacked(block_hash));
+    x[1] = encodeUint(round_num);
+    x[2] = encodeUint(uint(number));
+
+    bytes[] memory y = new bytes[](2);
+    y[0] = encodeList(x);
+    y[1] = encodeUint(gap_num);
+    signHash = keccak256(encodeList(y));
+  }
+
+  function toBytes32(bytes memory data) internal pure returns (bytes32 res) {
+    assembly {
+      res := mload(add(data, 32))
+    }
+  }
+
+  /*
+  * @param item RLP encoded bytes
+  */
+  function toRlpItem(bytes memory item) internal pure returns (RLPItem memory) {
```

Comment:
Starting from here, are these functions from some other source? If so, please give the reference. Then we can check whether the origin could have a bug or keep the code up to date.